vm/src/org/mozartoz/truffle/nodes/call/SelfTailCallCatcherNode.java

```diff
@@ -0,0 +1,81 @@
+package org.mozartoz.truffle.nodes.call;
+
+import org.mozartoz.truffle.nodes.OzNode;
+import org.mozartoz.truffle.runtime.OzBacktrace;
+import org.mozartoz.truffle.runtime.SelfTailCallException;
+
+import com.oracle.truffle.api.Truffle;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.LoopNode;
+import com.oracle.truffle.api.nodes.RepeatingNode;
+
+public class SelfTailCallCatcherNode extends OzNode {
+
+	OzNode body;
+	@Child LoopNode loopNode;
+
+	public SelfTailCallCatcherNode(OzNode body) {
+		this.body = body;
+		this.loopNode = Truffle.getRuntime().createLoopNode(new SelfTailCallLoopNode(body));
+	}
+
+	@Override
+	public Object execute(VirtualFrame frame) {
+		loopNode.executeLoop(frame);
+		return unit;
+	}
+
+	public static class SelfTailCallLoopNode extends OzNode implements RepeatingNode {
+
+		@Child OzNode body;
+
+		public SelfTailCallLoopNode(OzNode body) {
+			this.body = body;
+		}
+
+		@Override
+		public boolean executeRepeating(VirtualFrame frame) {
+			try {
+				body.execute(frame);
+				return false;
+			} catch (SelfTailCallException e) {
+				return true;
+			}
+		}
+
+		@Override
+		public Object execute(VirtualFrame frame) {
+			throw new UnsupportedOperationException();
+		}
+
+		@Override
+		public String toString() {
+			return OzBacktrace.formatNode(this);
+		}
+
+	}
+
+	/**
+	 * Same semantics, without OSR
+	 */
+	public static class NoOSR extends OzNode {
+
+		@Child OzNode body;
+
+		public NoOSR(OzNode body) {
+			this.body = body;
+		}
+
+		@Override
+		public Object execute(VirtualFrame frame) {
+			while (true) {
+				try {
+					return body.execute(frame);
+				} catch (SelfTailCallException tailCall) {
```

Comment:
In general you might want 2 BranchProfile for both cases here, but I guess in this case it's not necessary since Graal will see whether there is a path throwing an exception or not. OTOH, it could help partial evaluation to be faster. Leave like this for now.